File: crates/runtime/src/js/node/child_process.js
import EventEmitter from "ext:utoo_rt_ext/node/events";
import { Buffer } from "ext:utoo_rt_ext/node/buffer";

const ops = Deno.core.ops;
const asyncOps = Deno.core.ops;

class ChildProcess extends EventEmitter {
  constructor() {
    super();
    this.stdin = null;
    this.stdout = null;
    this.stderr = null;
    this.pid = 0;
    this.connected = false;
    this.killed = false;
    this.exitCode = null;
    this.signalCode = null;
    this._rid = null;
    this.channel = null;
  }
  kill(signal) {
    if (this._rid != null) {
      try { ops.op_spawn_kill(this._rid, typeof signal === "number" ? signal : 15); } catch(e) {}
    }
    this.killed = true;
    return true;
  }
  disconnect() { this.connected = false; }
  send(msg, handle, options, cb) {
    if (typeof cb === "function") cb(new Error("IPC not supported"));
    return false;
  }
  ref() { return this; }
  unref() { return this; }
}

function _buildEnvPairs(envObj) {
  if (!envObj) return [];
  // Ensure all keys and values are strings (serde expects Vec<(String, String)>)
  return Object.entries(envObj).filter(([k, v]) => v != null).map(([k, v]) => [String(k), String(v)]);
}

function spawn(cmd, args, options) {
  if (!Array.isArray(args)) { options = args; args = []; }
  options = options || {};

  const envPairs = _buildEnvPairs(options.env || process.env);
  const cwd = options.cwd || "";

  // Handle shell option
  let actualCmd = cmd;
  let actualArgs = args || [];
  if (options.shell) {
    const shellBin = typeof options.shell === "string" ? options.shell : "/bin/sh";
    const fullCmd = [cmd, ...(args || [])].join(" ");
    actualCmd = shellBin;
    actualArgs = ["-c", fullCmd];
  }

  // stdio config
  let stdio = ["pipe", "pipe", "pipe"];
  if (options.stdio) {
    if (typeof options.stdio === "string") {
      stdio = [options.stdio, options.stdio, options.stdio];
    } else if (Array.isArray(options.stdio)) {
      stdio = options.stdio.slice(0, 3).map(s => s || "pipe");
      while (stdio.length < 3) stdio.push("pipe");
    }
  }
  // Map to strings
  stdio = stdio.map(s => {
    if (s === "ignore") return "null";
    if (s === null || s === undefined) return "pipe";
    return String(s);
  });

  const cp = new ChildProcess();

  try {
    const result = ops.op_spawn(actualCmd, actualArgs, stdio, cwd, envPairs);
    cp._rid = result.rid;
    cp.pid = result.pid;

    if (result.stdin_rid != null) {
      cp.stdin = new StdioWritable(result.stdin_rid);
    }
    if (result.stdout_rid != null) {
      cp.stdout = new StdioReadable(result.stdout_rid);
    }
    if (result.stderr_rid != null) {
      cp.stderr = new StdioReadable(result.stderr_rid);
    }

    // Wait for exit asynchronously
    asyncOps.op_spawn_wait(result.rid).then(waitResult => {
      cp.exitCode = waitResult.code;
      cp.emit("exit", waitResult.code, null);
      cp.emit("close", waitResult.code, null);
    }).catch(e => {
      cp.emit("error", e);
    });
  } catch (e) {
    process.nextTick(() => cp.emit("error", e));
  }

  return cp;
}

// Simple readable stream for child stdout/stderr
class StdioReadable extends EventEmitter {
  constructor(rid) {
    super();
    this._rid = rid;
    this.readable = true;
    this._reading = false;
    this._ended = false;
    // Start reading on next tick if there are data listeners
    const self = this;
    const origOn = this.on.bind(this);
    this.on = function(event, listener) {
      origOn(event, listener);
      if (event === "data" && !self._reading) {
        self._startReading();
      }
      return self;
    };
    this.addListener = this.on;
  }
  _startReading() {
    if (this._reading) return;
    this._reading = true;
    const self = this;
    const opName = "op_spawn_stdout_read"; // works for both stdout and stderr via different rids
    (async function readLoop() {
      while (!self._ended) {
        try {
          const data = await asyncOps.op_spawn_stdout_read(self._rid, 65536);
          if (!data || data.length === 0) {
            self._ended = true;
            self.emit("end");
            break;
          }
          self.emit("data", Buffer.from(data));
        } catch (e) {
          self._ended = true;
          if (!e.message.includes("closed")) {
            self.emit("error", e);
          }
          self.emit("end");
          break;
        }
      }
    })();
  }
  pipe(dest) {
    this.on("data", chunk => dest.write(chunk));
    this.on("end", () => { if (dest !== process.stdout && dest !== process.stderr) dest.end(); });
    return dest;
  }
  setEncoding(enc) { this._encoding = enc; return this; }
  destroy() { this._ended = true; }
}

// Simple writable for child stdin
class StdioWritable extends EventEmitter {
  constructor(rid) {
    super();
    this._rid = rid;
    this.writable = true;
  }
  write(data, encoding, cb) {
    if (typeof encoding === "function") { cb = encoding; encoding = undefined; }
    const buf = typeof data === "string" ? Buffer.from(data, encoding || "utf8") : data;
    asyncOps.op_spawn_stdin_write(this._rid, buf).then(() => {
      if (cb) cb(null);
    }).catch(e => {
      if (cb) cb(e);
    });
    return true;
  }
  end(data, encoding, cb) {
    if (data) {
      this.write(data, encoding, () => {
        try { ops.op_spawn_stdin_close(this._rid); } catch(e) {}
        if (cb) cb();
      });
    } else {
      try { ops.op_spawn_stdin_close(this._rid); } catch(e) {}
      if (cb) cb();
    }
  }
  destroy() {
    try { ops.op_spawn_stdin_close(this._rid); } catch(e) {}
  }
}

function execSync(cmd, options) {
  options = options || {};
  const envPairs = _buildEnvPairs(options.env || process.env);
  const cwd = options.cwd || "";
  const shell = typeof options.shell === "string" ? options.shell : "/bin/sh";

  const result = ops.op_exec_sync(cmd, [], cwd, envPairs, shell);

  if (result.status !== 0 && !options.ignoreExitCode) {
    const err = new Error("Command failed: " + cmd);
    err.status = result.status;
    err.stdout = Buffer.from(result.stdout);
    err.stderr = Buffer.from(result.stderr);
    err.output = [null, err.stdout, err.stderr];
    throw err;
  }

  if (options.encoding === "buffer" || !options.encoding) {
    return Buffer.from(result.stdout);
  }
  return Buffer.from(result.stdout).toString(options.encoding || "utf8");
}

function exec(cmd, options, cb) {
  if (typeof options === "function") { cb = options; options = {}; }
  options = options || {};

  const cp = spawn("/bin/sh", ["-c", cmd], {
    ...options,
    stdio: ["pipe", "pipe", "pipe"],
  });

  let stdout = "";
  let stderr = "";

  if (cp.stdout) cp.stdout.on("data", d => { stdout += d.toString(); });
  if (cp.stderr) cp.stderr.on("data", d => { stderr += d.toString(); });

  cp.on("error", e => { if (cb) cb(e, stdout, stderr); });
  cp.on("close", (code) => {
    if (cb) {
      if (code !== 0) {
        const err = new Error("Command failed: " + cmd);
        err.code = code;
        cb(err, stdout, stderr);
      } else {
        cb(null, stdout, stderr);
      }
    }
  });

  return cp;
}

function execFile(file, args, options, cb) {
  if (typeof args === "function") { cb = args; args = []; options = {}; }
  else if (typeof options === "function") { cb = options; options = {}; }
  options = options || {};

  const cp = spawn(file, args || [], {
    ...options,
    stdio: ["pipe", "pipe", "pipe"],
  });

  let stdout = "";
  let stderr = "";

  if (cp.stdout) cp.stdout.on("data", d => { stdout += d.toString(); });
  if (cp.stderr) cp.stderr.on("data", d => { stderr += d.toString(); });

  cp.on("error", e => { if (cb) cb(e, stdout, stderr); });
  cp.on("close", (code) => {
    if (cb) {
      if (code !== 0) {
        const err = new Error("Command failed: " + file);
        err.code = code;
        cb(err, stdout, stderr);
      } else {
        cb(null, stdout, stderr);
      }
    }
  });

  return cp;
}

function execFileSync(file, args, options) {
  options = options || {};
  const envPairs = _buildEnvPairs(options.env || process.env);
  const cwd = options.cwd || "";

  const result = ops.op_exec_sync(file, args || [], cwd, envPairs, "");

  if (result.status !== 0) {
    const err = new Error("Command failed: " + file);
    err.status = result.status;
    err.stdout = Buffer.from(result.stdout);
    err.stderr = Buffer.from(result.stderr);
    throw err;
  }

  if (options.encoding === "buffer" || !options.encoding) {
    return Buffer.from(result.stdout);
  }
  return Buffer.from(result.stdout).toString(options.encoding || "utf8");
}

function spawnSync(cmd, args, options) {
  if (!Array.isArray(args)) { options = args; args = []; }
  options = options || {};
  const envPairs = _buildEnvPairs(options.env || process.env);
  const cwd = options.cwd || "";
  const shell = options.shell ? (typeof options.shell === "string" ? options.shell : "/bin/sh") : "";

  let actualCmd = cmd;
  let actualArgs = args || [];
  if (shell) {
    actualCmd = shell;
    actualArgs = ["-c", [cmd, ...(args || [])].join(" ")];
  }

  const result = ops.op_exec_sync(actualCmd, actualArgs, cwd, envPairs, "");

  return {
    status: result.status,
    stdout: Buffer.from(result.stdout),
    stderr: Buffer.from(result.stderr),
    output: [null, Buffer.from(result.stdout), Buffer.from(result.stderr)],
    pid: 0,
    signal: null,
    error: result.status !== 0 ? new Error("Command failed") : undefined,
  };
}

function fork(modulePath, args, options) {
  if (!Array.isArray(args)) { options = args; args = []; }
  options = options || {};

  const execPath = options.execPath || process.execPath;
  const execArgv = options.execArgv || process.execArgv || [];
  const allArgs = [...execArgv, modulePath, ...(args || [])];

  // Build env pairs and stdio config (same logic as spawn)
  const envPairs = _buildEnvPairs(options.env || process.env);
  const cwd = String(options.cwd || "");

  let stdio = options.stdio || ["pipe", "inherit", "inherit"];
  if (typeof stdio === "string") {
    stdio = [stdio, stdio, stdio];
  } else if (Array.isArray(stdio)) {
    stdio = stdio.slice(0, 3).map(s => s || "pipe");
    while (stdio.length < 3) stdio.push("pipe");
  }
  stdio = stdio.map(s => {
    if (s === "ignore") return "null";
    if (s === "ipc") return "pipe"; // IPC is handled separately
    if (s === null || s === undefined) return "pipe";
    return String(s);
  });

  const cp = new ChildProcess();

  try {
    // Use op_fork_spawn which creates a socketpair for IPC
    const result = ops.op_fork_spawn(String(execPath), allArgs.map(String), stdio, cwd, envPairs);
    cp._rid = result.rid;
    cp.pid = result.pid;
    cp._ipcRid = result.ipc_rid;
    cp.connected = true;
    cp.channel = { ref() {}, unref() {} };

    if (result.stdin_rid != null) {
      cp.stdin = new StdioWritable(result.stdin_rid);
    }
    if (result.stdout_rid != null) {
      cp.stdout = new StdioReadable(result.stdout_rid);
    }
    if (result.stderr_rid != null) {
      cp.stderr = new StdioReadable(result.stderr_rid);
    }

    // Set up cp.send() for parent -> child IPC
    cp.send = function(message, handle, options2, callback) {
      if (typeof handle === "function") { callback = handle; handle = undefined; options2 = undefined; }
      if (typeof options2 === "function") { callback = options2; options2 = undefined; }
      if (!cp.connected || cp._ipcRid == null) {
        if (typeof callback === "function") callback(new Error("IPC channel closed"));
        return false;
      }
      const json = JSON.stringify(message);
      asyncOps.op_ipc_stream_write(cp._ipcRid, json).then(() => {
        if (typeof callback === "function") callback(null);
      }).catch(e => {
        if (typeof callback === "function") callback(e);
      });
      return true;
    };

    // Start IPC read loop: parent reads messages from child
    (async function ipcReadLoop() {
      while (cp.connected && cp._ipcRid != null) {
        try {
          const line = await asyncOps.op_ipc_stream_read_line(cp._ipcRid);
          if (!line) {
            // EOF
            cp.connected = false;
            cp.channel = null;
            cp.emit("disconnect");
            break;
          }
          try {
            const msg = JSON.parse(line);
            cp.emit("message", msg);
          } catch (e) {
            // Non-JSON line, ignore
          }
        } catch (e) {
          cp.connected = false;
          cp.channel = null;
          cp.emit("disconnect");
          break;
        }
      }
    })();

    // Wait for exit asynchronously
    asyncOps.op_spawn_wait(result.rid).then(waitResult => {
      cp.exitCode = waitResult.code;
      cp.connected = false;
      cp.channel = null;
      cp.emit("exit", waitResult.code, null);
      cp.emit("close", waitResult.code, null);
    }).catch(e => {
      cp.emit("error", e);
    });
  } catch (e) {
    process.nextTick(() => cp.emit("error", e));
  }

  return cp;
}

const child_process = {
  ChildProcess,
  exec,
  execSync,
  execFile,
  execFileSync,
  spawn,
  spawnSync,
  fork,
};
child_process.default = child_process;

export default child_process;
export { ChildProcess, exec, execSync, execFile, execFileSync, spawn, spawnSync, fork };

File: crates/runtime/src/ops/child_process.rs
use std::borrow::Cow;
use std::cell::RefCell;
use std::rc::Rc;

use deno_core::{op2, JsBuffer, OpState, Resource, ResourceId};
use deno_error::JsErrorBox;
use serde::Serialize;
use tokio::io::{AsyncBufReadExt, AsyncReadExt, AsyncWriteExt};
use tokio::process::{Child, Command};

struct ChildProcessResource {
    child: RefCell<Option<Child>>,
    stdin_rid: Option<ResourceId>,
    stdout_rid: Option<ResourceId>,
    stderr_rid: Option<ResourceId>,
}

impl Resource for ChildProcessResource {
    fn name(&self) -> Cow<'_, str> {
        "childProcess".into()
    }
}

struct ChildStdinResource {
    inner: RefCell<Option<tokio::process::ChildStdin>>,
}

impl Resource for ChildStdinResource {
    fn name(&self) -> Cow<'_, str> {
        "childStdin".into()
    }
}

struct ChildStdoutResource {
    inner: RefCell<Option<tokio::process::ChildStdout>>,
}

impl Resource for ChildStdoutResource {
    fn name(&self) -> Cow<'_, str> {
        "childStdout".into()
    }
}

struct ChildStderrResource {
    inner: RefCell<Option<tokio::process::ChildStderr>>,
}

impl Resource for ChildStderrResource {
    fn name(&self) -> Cow<'_, str> {
        "childStderr".into()
    }
}

#[derive(Serialize)]
pub struct SpawnResult {
    pub rid: ResourceId,
    pub pid: u32,
    pub stdin_rid: Option<ResourceId>,
    pub stdout_rid: Option<ResourceId>,
    pub stderr_rid: Option<ResourceId>,
}

#[derive(Serialize)]
pub struct ForkSpawnResult {
    pub rid: ResourceId,
    pub pid: u32,
    pub stdin_rid: Option<ResourceId>,
    pub stdout_rid: Option<ResourceId>,
    pub stderr_rid: Option<ResourceId>,
    pub ipc_rid: ResourceId,
}

#[derive(Serialize)]
pub struct WaitResult {
    pub code: Option<i32>,
    pub success: bool,
}

/// Resource wrapping the parent end of an IPC socketpair.
struct IpcStreamResource {
    reader: RefCell<tokio::io::BufReader<tokio::io::ReadHalf<tokio::net::UnixStream>>>,
    writer: RefCell<tokio::io::WriteHalf<tokio::net::UnixStream>>,
}

impl Resource for IpcStreamResource {
    fn name(&self) -> Cow<'_, str> {
        "ipcStream".into()
    }
}

/// Spawn a child process.
/// `stdio` is a 3-element array: [stdin, stdout, stderr]
/// Each element: "pipe" | "inherit" | "null"
#[op2]
#[serde]
pub fn op_spawn(
    state: &mut OpState,
    #[string] cmd: &str,
    #[serde] args: Vec<String>,
    #[serde] stdio: Vec<String>,
    #[string] cwd: &str,
    #[serde] env: Vec<(String, String)>,
) -> Result<SpawnResult, JsErrorBox> {
    let mut command = Command::new(cmd);
    command.args(&args);

    if !cwd.is_empty() {
        command.current_dir(cwd);
    }

    for (k, v) in &env {
        command.env(k, v);
    }

    // Configure stdio
    let stdin_cfg = stdio.first().map(|s| s.as_str()).unwrap_or("pipe");
    let stdout_cfg = stdio.get(1).map(|s| s.as_str()).unwrap_or("pipe");
    let stderr_cfg = stdio.get(2).map(|s| s.as_str()).unwrap_or("pipe");

    command.stdin(match stdin_cfg {
        "pipe" => std::process::Stdio::piped(),
        "inherit" => std::process::Stdio::inherit(),
        _ => std::process::Stdio::null(),
    });
    command.stdout(match stdout_cfg {
        "pipe" => std::process::Stdio::piped(),
        "inherit" => std::process::Stdio::inherit(),
        _ => std::process::Stdio::null(),
    });
    command.stderr(match stderr_cfg {
        "pipe" => std::process::Stdio::piped(),
        "inherit" => std::process::Stdio::inherit(),
        _ => std::process::Stdio::null(),
    });

    let mut child = command
        .spawn()
        .map_err(|e| JsErrorBox::generic(format!("Failed to spawn '{}': {}", cmd, e)))?;

    let pid = child.id().unwrap_or(0);

    // Extract stdio handles and register as resources
    let stdin_rid = child.stdin.take().map(|s| {
        state.resource_table.add(ChildStdinResource {
            inner: RefCell::new(Some(s)),
        })
    });
    let stdout_rid = child.stdout.take().map(|s| {
        state.resource_table.add(ChildStdoutResource {
            inner: RefCell::new(Some(s)),
        })
    });
    let stderr_rid = child.stderr.take().map(|s| {
        state.resource_table.add(ChildStderrResource {
            inner: RefCell::new(Some(s)),
        })
    });

    let rid = state.resource_table.add(ChildProcessResource {
        child: RefCell::new(Some(child)),
        stdin_rid,
        stdout_rid,
        stderr_rid,
    });

    Ok(SpawnResult {
        rid,
        pid,
        stdin_rid,
        stdout_rid,
        stderr_rid,
    })
}

/// Wait for a child process to exit.
#[op2]
#[serde]
pub async fn op_spawn_wait(
    state: Rc<RefCell<OpState>>,
    #[smi] rid: ResourceId,
) -> Result<WaitResult, JsErrorBox> {
    let resource = state
        .borrow()
        .resource_table
        .get::<ChildProcessResource>(rid)
        .map_err(|e| JsErrorBox::generic(e.to_string()))?;

    let mut child_opt = resource.child.borrow_mut();
    let child = child_opt
        .as_mut()
        .ok_or_else(|| JsErrorBox::generic("Child process already consumed"))?;

    let status = child
        .wait()
        .await
        .map_err(|e| JsErrorBox::generic(e.to_string()))?;

    Ok(WaitResult {
        code: status.code(),
        success: status.success(),
    })
}

/// Write to child stdin.
#[op2]
#[smi]
pub async fn op_spawn_stdin_write(
    state: Rc<RefCell<OpState>>,
    #[smi] rid: ResourceId,
    #[buffer] data: JsBuffer,
) -> Result<u32, JsErrorBox> {
    let resource = state
        .borrow()
        .resource_table
        .get::<ChildStdinResource>(rid)
        .map_err(|e| JsErrorBox::generic(e.to_string()))?;

    let mut inner = resource.inner.borrow_mut();
    let stdin = inner
        .as_mut()
        .ok_or_else(|| JsErrorBox::generic("Stdin already closed"))?;

    let n = stdin
        .write(data.as_ref())
        .await
        .map_err(|e| JsErrorBox::generic(e.to_string()))?;
    Ok(n as u32)
}

/// Close child stdin (signals EOF to the child).
#[op2(fast)]
pub fn op_spawn_stdin_close(
    state: &mut OpState,
    #[smi] rid: ResourceId,
) -> Result<(), JsErrorBox> {
    let resource = state
        .resource_table
        .get::<ChildStdinResource>(rid)
        .map_err(|e| JsErrorBox::generic(e.to_string()))?;
    resource.inner.borrow_mut().take(); // Drop the stdin handle
    Ok(())
}

/// Read from child stdout.
#[op2]
#[serde]
pub async fn op_spawn_stdout_read(
    state: Rc<RefCell<OpState>>,
    #[smi] rid: ResourceId,
    #[smi] len: u32,
) -> Result<Vec<u8>, JsErrorBox> {
    let resource = state
        .borrow()
        .resource_table
        .get::<ChildStdoutResource>(rid)
        .map_err(|e| JsErrorBox::generic(e.to_string()))?;

    let mut inner = resource.inner.borrow_mut();
    let stdout = inner
        .as_mut()
        .ok_or_else(|| JsErrorBox::generic("Stdout already closed"))?;

    let mut buf = vec![0u8; len as usize];
    let n = stdout
        .read(&mut buf)
        .await
        .map_err(|e| JsErrorBox::generic(e.to_string()))?;
    buf.truncate(n);
    Ok(buf)
}

/// Read from child stderr.
#[op2]
#[serde]
pub async fn op_spawn_stderr_read(
    state: Rc<RefCell<OpState>>,
    #[smi] rid: ResourceId,
    #[smi] len: u32,
) -> Result<Vec<u8>, JsErrorBox> {
    let resource = state
        .borrow()
        .resource_table
        .get::<ChildStderrResource>(rid)
        .map_err(|e| JsErrorBox::generic(e.to_string()))?;

    let mut inner = resource.inner.borrow_mut();
    let stderr = inner
        .as_mut()
        .ok_or_else(|| JsErrorBox::generic("Stderr already closed"))?;

    let mut buf = vec![0u8; len as usize];
    let n = stderr
        .read(&mut buf)
        .await
        .map_err(|e| JsErrorBox::generic(e.to_string()))?;
    buf.truncate(n);
    Ok(buf)
}

/// Kill a child process.
#[op2(fast)]
pub fn op_spawn_kill(
    state: &mut OpState,
    #[smi] rid: ResourceId,
    #[smi] signal: i32,
) -> Result<(), JsErrorBox> {
    let resource = state
        .resource_table
        .get::<ChildProcessResource>(rid)
        .map_err(|e| JsErrorBox::generic(e.to_string()))?;

    let mut child_opt = resource.child.borrow_mut();
    let child = child_opt
        .as_mut()
        .ok_or_else(|| JsErrorBox::generic("Child process already consumed"))?;

    // On Unix, send signal via libc. Default to SIGTERM.
    #[cfg(unix)]
    {
        let _ = signal; // We use the signal parameter
        if let Some(pid) = child.id() {
            let sig = if signal == 0 { 15 } else { signal }; // SIGTERM = 15
            unsafe {
                libc::kill(pid as libc::pid_t, sig);
            }
        }
    }

    #[cfg(not(unix))]
    {
        let _ = signal;
        child
            .start_kill()
            .map_err(|e| JsErrorBox::generic(e.to_string()))?;
    }

    Ok(())
}

/// Close/cleanup a child process resource.
#[op2(fast)]
pub fn op_spawn_close(state: &mut OpState, #[smi] rid: ResourceId) -> Result<(), JsErrorBox> {
    // Try to remove all related resources
    if let Ok(proc) = state.resource_table.take::<ChildProcessResource>(rid) {
        if let Some(stdin_rid) = proc.stdin_rid {
            let _ = state.resource_table.take::<ChildStdinResource>(stdin_rid);
        }
        if let Some(stdout_rid) = proc.stdout_rid {
            let _ = state.resource_table.take::<ChildStdoutResource>(stdout_rid);
        }
        if let Some(stderr_rid) = proc.stderr_rid {
            let _ = state.resource_table.take::<ChildStderrResource>(stderr_rid);
        }
    }
    Ok(())
}

/// Execute a command synchronously and return its stdout.
#[derive(Serialize)]
pub struct ExecSyncResult {
    pub stdout: Vec<u8>,
    pub stderr: Vec<u8>,
    pub status: i32,
}

#[op2]
#[serde]
pub fn op_exec_sync(
    #[string] cmd: &str,
    #[serde] args: Vec<String>,
    #[string] cwd: &str,
    #[serde] env: Vec<(String, String)>,
    #[string] shell: &str,
) -> Result<ExecSyncResult, JsErrorBox> {
    let mut command = if !shell.is_empty() {
        let mut c = std::process::Command::new(shell);
        c.arg("-c");
        let full = if args.is_empty() {
            cmd.to_string()
        } else {
            format!("{} {}", cmd, args.join(" "))
        };
        c.arg(full);
        c
    } else {
        let mut c = std::process::Command::new(cmd);
        c.args(&args);
        c
    };

    if !cwd.is_empty() {
        command.current_dir(cwd);
    }

    for (k, v) in &env {
        command.env(k, v);
    }

    let output = command
        .output()
        .map_err(|e| JsErrorBox::generic(format!("execSync failed: {}", e)))?;

    let status = output.status.code().unwrap_or(-1);

    Ok(ExecSyncResult {
        stdout: output.stdout,
        stderr: output.stderr,
        status,
    })
}

/// Spawn a child process with an IPC channel (Unix socketpair on fd 3).
/// This is the equivalent of Node.js `child_process.fork()`.
#[op2]
#[serde]
pub fn op_fork_spawn(
    state: &mut OpState,
    #[string] cmd: &str,
    #[serde] args: Vec<String>,
    #[serde] stdio: Vec<String>,
    #[string] cwd: &str,
    #[serde] env: Vec<(String, String)>,
) -> Result<ForkSpawnResult, JsErrorBox> {
    use std::os::unix::io::FromRawFd;
    #[allow(unused_imports)] // used by pre_exec() inside unsafe block
    use std::os::unix::process::CommandExt;

    // Create a Unix socketpair for IPC
    let mut fds: [libc::c_int; 2] = [0; 2];
    let ret = unsafe { libc::socketpair(libc::AF_UNIX, libc::SOCK_STREAM, 0, fds.as_mut_ptr()) };
    if ret != 0 {
        return Err(JsErrorBox::generic(format!(
            "socketpair failed: {}",
            std::io::Error::last_os_error()
        )));
    }
    let parent_fd = fds[0];
    let child_fd = fds[1];

    let mut command = Command::new(cmd);
    command.args(&args);

    if !cwd.is_empty() {
        command.current_dir(cwd);
    }

    for (k, v) in &env {
        command.env(k, v);
    }
    // Tell the child which fd is the IPC channel
    command.env("NODE_CHANNEL_FD", "3");

    // Configure stdio (same as op_spawn)
    let stdin_cfg = stdio.first().map(|s| s.as_str()).unwrap_or("pipe");
    let stdout_cfg = stdio.get(1).map(|s| s.as_str()).unwrap_or("pipe");
    let stderr_cfg = stdio.get(2).map(|s| s.as_str()).unwrap_or("pipe");

    command.stdin(match stdin_cfg {
        "pipe" => std::process::Stdio::piped(),
        "inherit" => std::process::Stdio::inherit(),
        _ => std::process::Stdio::null(),
    });
    command.stdout(match stdout_cfg {
        "pipe" => std::process::Stdio::piped(),
        "inherit" => std::process::Stdio::inherit(),
        _ => std::process::Stdio::null(),
    });
    command.stderr(match stderr_cfg {
        "pipe" => std::process::Stdio::piped(),
        "inherit" => std::process::Stdio::inherit(),
        _ => std::process::Stdio::null(),
    });

    // Before spawning, set up fd 3 in the child to be the child end of the socketpair
    unsafe {
        let cfd = child_fd;
        command.pre_exec(move || {
            // If child_fd is not already 3, dup it to 3
            if cfd != 3 {
                libc::dup2(cfd, 3);
                libc::close(cfd);
            }
            // Clear close-on-exec for fd 3 so it survives exec
            libc::fcntl(3, libc::F_SETFD, 0);
            Ok(())
        });
    }

    let mut child = command
        .spawn()
        .map_err(|e| JsErrorBox::generic(format!("Failed to fork-spawn '{}': {}", cmd, e)))?;

    // Close the child end of the socketpair in the parent
    unsafe {
        libc::close(child_fd);
    }

    let pid = child.id().unwrap_or(0);

    // Extract stdio handles
    let stdin_rid = child.stdin.take().map(|s| {
        state.resource_table.add(ChildStdinResource {
            inner: RefCell::new(Some(s)),
        })
    });
    let stdout_rid = child.stdout.take().map(|s| {
        state.resource_table.add(ChildStdoutResource {
            inner: RefCell::new(Some(s)),
        })
    });
    let stderr_rid = child.stderr.take().map(|s| {
        state.resource_table.add(ChildStderrResource {
            inner: RefCell::new(Some(s)),
        })
    });

    let rid = state.resource_table.add(ChildProcessResource {
        child: RefCell::new(Some(child)),
        stdin_rid,
        stdout_rid,
        stderr_rid,
    });

    // Wrap the parent end of the socketpair as a tokio UnixStream
    let parent_std = unsafe { std::os::unix::net::UnixStream::from_raw_fd(parent_fd) };
    parent_std
        .set_nonblocking(true)
        .map_err(|e| JsErrorBox::generic(format!("set_nonblocking: {}", e)))?;
    let parent_stream = tokio::net::UnixStream::from_std(parent_std)
        .map_err(|e| JsErrorBox::generic(format!("UnixStream::from_std: {}", e)))?;
    let (read_half, write_half) = tokio::io::split(parent_stream);

    let ipc_rid = state.resource_table.add(IpcStreamResource {
        reader: RefCell::new(tokio::io::BufReader::new(read_half)),
        writer: RefCell::new(write_half),
    });

    Ok(ForkSpawnResult {
        rid,
        pid,
        stdin_rid,
        stdout_rid,
        stderr_rid,
        ipc_rid,
    })
}

/// Write a message (JSON line) to the parent end of an IPC channel.
#[op2]
#[smi]
pub async fn op_ipc_stream_write(
    state: Rc<RefCell<OpState>>,
    #[smi] rid: ResourceId,
    #[string] data: String,
) -> Result<u32, JsErrorBox> {
    let resource = state
        .borrow()
        .resource_table
        .get::<IpcStreamResource>(rid)
        .map_err(|e| JsErrorBox::generic(e.to_string()))?;

    let msg = format!("{}\n", data);
    let mut writer = resource.writer.borrow_mut();
    writer
        .write_all(msg.as_bytes())
        .await
        .map_err(|e| JsErrorBox::generic(format!("IPC write: {}", e)))?;
    writer
        .flush()
        .await
        .map_err(|e| JsErrorBox::generic(format!("IPC flush: {}", e)))?;
    Ok(msg.len() as u32)
}

/// Read a newline-delimited message from the parent end of an IPC channel.
#[op2]
#[string]
pub async fn op_ipc_stream_read_line(
    state: Rc<RefCell<OpState>>,
    #[smi] rid: ResourceId,
) -> Result<String, JsErrorBox> {
    let resource = state
        .borrow()
        .resource_table
        .get::<IpcStreamResource>(rid)
        .map_err(|e| JsErrorBox::generic(e.to_string()))?;

    let mut reader = resource.reader.borrow_mut();
    let mut line = String::new();
    let n = reader
        .read_line(&mut line)
        .await
        .map_err(|e| JsErrorBox::generic(format!("IPC read: {}", e)))?;
    if n == 0 {
        return Ok(String::new()); // EOF
    }
    if line.ends_with('\n') {
        line.pop();
    }
    if line.ends_with('\r') {
        line.pop();
    }
    Ok(line)
}
